linera-views/src/journaling.rs

```diff
@@ -231,28 +250,77 @@ where
         Ok(())
     }
 
-    /// Writes blocks to the database and resolves them later.
+    /// Writes the content of `batch` to the journal as a succession of blocks that can be
+    /// interpreted later by `coherently_resolve_journal`.
+    ///
+    /// Starting with a batch of operations that is typically too large to be executed in
+    /// one go (see `is_fastpath_feasible()` below), the goal of this function is to split
+    /// the batch into smaller blocks so that `coherently_resolve_journal` respects the
+    /// constraints of the underlying key-value store (see analysis above).
+    ///
+    /// For efficiency reasons, we write as many blocks as possible in each "transaction"
+    /// batch, using one write-operation per block. Then we also update the journal header
+    /// with the final number of blocks.
+    ///
+    /// As a result, the constraints of the underlying database are respected if the
+    /// following conditions are met while a "transaction" batch is being built:
+    ///
+    /// (1) the number of blocks per transaction doesn't exceed `K::MAX_BATCH_SIZE`;
+    ///
+    /// (2) the total size of BCS-serialized blocks in a transaction doesn't exceed
+    /// `K::MAX_BATCH_TOTAL_SIZE - K::MAX_BATCH_SIZE * sizeof(block_key)`
+    ///
+    /// (3) the size of each BCS-serialized block doesn't exceed `K::MAX_VALUE_SIZE`.
+    ///
+    /// (4) `block_key` and `header_key` don't exceed `K::MAX_KEY_SIZE` and `bcs_header`
+    /// doesn't exceed `K::MAX_VALUE_SIZE`.
+    ///
+    /// NOTE:
+    /// * Since a block must contain at least one operation and M bytes of the
+    /// serialization overhead (typically M = 2 or 3 bytes of vector sizes), condition (3)
+    /// requires that each operation in the original batch satisfies:
+    ///     `sizeof(key) + sizeof(value) + M <= K::MAX_VALUE_SIZE`
+    ///
+    /// * Similarly, a transaction must contain at least one block so it is desirable that
+    /// the maximum size of a block insertion `1 + sizeof(block_key) + K::MAX_VALUE_SIZE`
+    /// plus M bytes of overhead doesn't exceed the threshold of condition (2).
     async fn write_journal(
         &self,
-        simplified_batch: K::Batch,
+        batch: K::Batch,
         base_key: &[u8],
     ) -> Result<JournalHeader, K::Error> {
-        let mut iter = simplified_batch.into_iter();
+        let header_key = get_journaling_key(base_key, KeyTag::Journal as u8, 0)?;
+        let header_key_len = header_key.len();
+        let header_len = bcs::serialized_size(&JournalHeader::default())?;
+        // Currently the same length.
+        let block_key_len = header_key.len();
+        // Each block in a transaction comes with a key.
+        let max_transaction_size = K::MAX_BATCH_TOTAL_SIZE - K::MAX_BATCH_SIZE * block_key_len;
```

Comment:
I actually can get an overflow here for `ScyllaDb`.
The thing is that for `ScyllaDb` the size of the batch is limited to 16M, but the number of transactions inside is not upper bound. So `MAX_BATCH_TOTAL_SIZE=16*1024*1024` and `MAX_BATCH_SIZE = USIZE::MAX`.